### PR TITLE
Bind lambda in `yield return` despite errors

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/68027

The problem was that `BindYieldReturnStatement` gave up too easily on trying to convert the lambda. Without conversion, the lambda state didn't account for the target delegate type, which provides the parameter types for the lambda. So the lookup result for identifier `s` (which finds a candidate for type in local declaration `s await`) lacked a type.

In fixing this, I noticed that the binding diagnostics for the `yield return` and the `return` scenarios were different. 
The `ERR_BadSKknown` and `ERR_IllegalStatement` binding errors were now reported in the `yield return` scenario, but not in the `return` scenario. 
In `BindYieldReturn`, we use `GenerateConversionForAssignment` which collects the conversion diagnostics from `GenerateImplicitConversionError` unless `expression.HasAnyErrors && expression.Kind != BoundKind.UnboundLambda`.
In `BindReturn`, we use `CreateReturnConversion`, which skipped reporting from `GenerateImplicitConversionError` if `argument.HasAnyErrors`.
So I made an adjustment to the `return` binding logic to also report the diagnostics from converting/binding the lambda (`BoundKind.UnboundLambda`).

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 1]
```diff
@@ -239,40 +239,53 @@ private BoundStatement BindYieldReturnStatement(YieldStatementSyntax node, Bindi
...
-            }
+            bool hasErrors = false;
 
             // NOTE: it's possible that more than one of these conditions is satisfied and that
             // we won't report the syntactically innermost.  However, dev11 appears to check
             // them in this order, regardless of syntactic nesting (StatementBinder::bindYield).
             if (this.Flags.Includes(BinderFlags.InFinallyBlock))
             {
                 Error(diagnostics, ErrorCode.ERR_BadYieldInFinally, node.YieldKeyword);
+                hasErrors = true;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:250]: >hasErrors = true; [](http://example.com/codeflow?start=16&length=17)

Why would we want to disable conversion binding for this and the other modified code paths. It feels to me that the conversion checks and the "unsupported" context checks are orthogonal, therefore cannot be considered as cascading for one another

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 2]
```diff
@@ -239,40 +239,53 @@ private BoundStatement BindYieldReturnStatement(YieldStatementSyntax node, Bindi
...
             }
             else if (InUnsafeRegion && Compilation.IsFeatureEnabled(MessageID.IDS_FeatureRefUnsafeInIteratorAsync))
             {
                 Error(diagnostics, ErrorCode.ERR_BadYieldInUnsafe, node.YieldKeyword);
+                hasErrors = true;
             }
 
             CheckRequiredLangVersionForIteratorMethods(node, diagnostics);
+
+            if (argument != null)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:275]: >argument != null [](http://example.com/codeflow?start=16&length=16)

Is this condition ever `false`?

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 3]
```diff
@@ -3154,7 +3167,7 @@ internal BoundExpression CreateReturnConversion(
 
             diagnostics.Add(syntax, useSiteInfo);
 
-            if (!argument.HasAnyErrors)
+            if (!argument.HasAnyErrors || argument.Kind == BoundKind.UnboundLambda)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:3170]: >!argument.HasAnyErrors || argument.Kind == BoundKind.UnboundLambda [](http://example.com/codeflow?start=16&length=66)

It looks like according to PR description this change is motivated by consistency with `BindYieldReturnStatement`. However, the conditions do not look the same to me. My preference would be to not touch this code, or use the same condition. Specifically ```!(argument.HasErrors || ((object)argument.Type != null && argument.Type.IsErrorType()))```

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs [hunk 4]
```diff
@@ -8672,5 +8673,154 @@ static async System.Threading.Tasks.Task Main()
...
+        {
+            await Task.CompletedTask;
+            throw new NotImplementedException();
+        };
+    }
+}
+""";
+            var comp = CreateCompilation(src, targetFramework: TargetFramework.Net80);
+            comp.VerifyDiagnostics(
+                // (7,63): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs:8699]: >warning CS1998 [](http://example.com/codeflow?start=27&length=14)

Consider suppressing this warning if it is not a goal of the test to verify its presence

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: I woke up this morning with the same thought, but you'd already reviewed. Thanks

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 6)

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**